Sources/Data/Attributes.swift

```diff
@@ -0,0 +1,299 @@
+//
+//  Attributes.swift
+//  RoverData
+//
+//  Created by Sean Rucker on 2018-07-24.
+//  Copyright © 2018 Rover Labs Inc. All rights reserved.
+//
+
+import Foundation
+import os
+
+/// Wraps a [String: Any], but enables use of both NSCoding and Codable.  Enforces the
+class Attributes: NSObject, NSCoding, Codable {
+    var rawValue: [String: Any]
+    
+    public init(_ dictionary: [String:Any]) {
+        // to get an implicit pass of our validation logic, render it to our internal NS attributes, which as a side-effect can emit an error.
+        let _ = dictionary.asAttributesForNsDictionary
+        self.rawValue = dictionary
+    }
+
+    //
+    // MARK: NSCoding
+    //
+    
+    public func encode(with aCoder: NSCoder) {
+        let nsDictionary = rawValue.attributes
+        aCoder.encode(nsDictionary)
+    }
+    
+    public required init?(coder aDecoder: NSCoder) {
+        guard let nsDictionary = aDecoder.decodeObject() as? NSDictionary else {
+            return nil
+        }
+        rawValue = nsDictionary.attributes
+        super.init()
+    }
+    
+    //
+    // MARK: Codable
+    //
+    
+    /// This implementation of CodingKey allows for handling data without strongly and statically typed keys with Codable.
+    struct DynamicCodingKeys: CodingKey {
+        var stringValue: String
+        
+        init(stringValue: String) {
+            self.stringValue = stringValue
+        }
+        
+        var intValue: Int? {
+            return nil
+        }
+        
+        init?(intValue: Int) {
+            return nil
+        }
+    }
+    
+    required public init(from decoder: Decoder) throws {
+        func fromKeyedDecoder(_ container: KeyedDecodingContainer<Attributes.DynamicCodingKeys>) throws -> [String:Any] {
+            var assembledHash = [String: Any]()
+
+            try container.allKeys.forEach { key in
+                let keyString = key.stringValue
+                // TODO: Verify key is valid (matches regex)
+                
+                if let value = try? container.decode(Bool.self, forKey: key) {
+                    assembledHash[keyString] = value
+                    return
+                }
+                
+                if let value = try? container.decode(Int.self, forKey: key) {
+                    assembledHash[keyString] = value
+                    return
+                }
+                
+                if let value = try? container.decode(String.self, forKey: key) {
+                    assembledHash[keyString] = value
+                    return
+                }
+                
+                if (try? container.decodeNil(forKey: key)) == true {
+                    assembledHash[keyString] = nil
```

Comment:
[String: Any] can't contain `nil` values. Setting the value to `nil` simply omits it from the Dictionary and is therefore a no-op.